File: src/865. Smallest Subtree with all the Deepest Nodes/main.ts
// Solved by @kitanoyoru
// https://leetcode.com/problems/smallest-subtree-with-all-the-deepest-nodes

class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val
    this.left = left === undefined ? null : left
    this.right = right === undefined ? null : right
  }
}

const getHighestLevel = (node: TreeNode | null): number => {
  if (node == null) {
    return 0
  }

  return 1 + Math.max(getHighestLevel(node.left), getHighestLevel(node.right))
}

const dfs = (node: TreeNode | null): TreeNode | null => {
  let leftHeight = getHighestLevel(node!.left),
    rightHeight = getHighestLevel(node!.right)

  if (leftHeight > rightHeight) {
    return dfs(node!.left)
  } else if (leftHeight < rightHeight) {
    return dfs(node!.right)
  } else {
    return node
  }
}

const subtreeWithAllDeepest = (root: TreeNode | null): TreeNode | null => {
  return dfs(root)
}

export {}

File: src/865. Smallest Subtree with all the Deepest Nodes/main.go
// Solved by @kitanoyoru
// https://leetcode.com/problems/smallest-subtree-with-all-the-deepest-nodes/

package main

type TreeNode struct {
  Val int
  Left *TreeNode
  Right *TreeNode
}

func max(a, b int) int {
  if a < b {
    return b
  }

  return a
}

func getDeepestLevel(node *TreeNode) int {
  if node == nil {
    return 0;
  }

  return 1 + max(getDeepestLevel(node.Left), getDeepestLevel(node.Right)) 
}

func dfs(node *TreeNode) *TreeNode {
  leftHeight, rightHeight := getDeepestLevel(node.Left), getDeepestLevel(node.Right)
  
  if leftHeight > rightHeight {
    return dfs(node.Left)
  } else if leftHeight < rightHeight {
    return dfs(node.Right)
  } else {
    return node
  }
}

func subtreeWithAllDeepest(root *TreeNode) *TreeNode {
  return dfs(root) 
}
